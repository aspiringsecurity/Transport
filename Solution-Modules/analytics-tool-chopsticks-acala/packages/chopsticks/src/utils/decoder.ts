import '@polkadot/types-codec'
import { Block } from '../blockchain/block'
import { DecoratedMeta } from '@polkadot/types/metadata/decorate/types'
import { HexString } from '@polkadot/util/types'
import { StorageEntry } from '@polkadot/types/primitive/types'
import { StorageKey } from '@polkadot/types'
import { blake2AsHex } from '@polkadot/util-crypto'
import { create } from 'jsondiffpatch'
import { hexToU8a, u8aToHex } from '@polkadot/util'
import _ from 'lodash'

const diffPatcher = create({
  array: { detectMove: false },
  textDiff: { minLength: Number.MAX_VALUE }, // skip text diff
})

const _CACHE: Record<string, Record<HexString, StorageEntry>> = {}

const getCache = (uid: string): Record<HexString, StorageEntry> => {
  if (!_CACHE[uid]) {
    _CACHE[uid] = {}
  }
  return _CACHE[uid]
}

const getStorageEntry = (meta: DecoratedMeta, block: Block, key: HexString) => {
  const cache = getCache(block.chain.uid)
  for (const [prefix, storageEntry] of Object.entries(cache)) {
    if (key.startsWith(prefix)) return storageEntry
  }
  for (const module of Object.values(meta.query)) {
    for (const storage of Object.values(module)) {
      const keyPrefix = u8aToHex(storage.keyPrefix())
      if (key.startsWith(keyPrefix)) {
        cache[keyPrefix] = storage
        return storage
      }
    }
  }
  return undefined
}

export const decodeKey = (
  meta: DecoratedMeta,
  block: Block,
  key: HexString
): { storage?: StorageEntry; decodedKey?: StorageKey } => {
  const storage = getStorageEntry(meta, block, key)
  const decodedKey = meta.registry.createType('StorageKey', key)
  if (storage) {
    decodedKey.setMeta(storage.meta)
    return { storage, decodedKey }
  }
  return {}
}

export const decodeKeyValue = (meta: DecoratedMeta, block: Block, key: HexString, value?: HexString | null) => {
  const { storage, decodedKey } = decodeKey(meta, block, key)

  if (!storage || !decodedKey) {
    return { [key]: value }
  }

  const decodeValue = () => {
    if (!value) return null
    if (storage.section === 'substrate' && storage.method === 'code') {
      return `:code blake2_256 ${blake2AsHex(value, 256)} (${hexToU8a(value).length} bytes)`
    }
    return meta.registry.createType(decodedKey.outputType, hexToU8a(value)).toHuman()
  }

  switch (decodedKey.args.length) {
    case 2: {
      return {
        [storage.section]: {
          [storage.method]: {
            [decodedKey.args[0].toString()]: {
              [decodedKey.args[1].toString()]: decodeValue(),
            },
          },
        },
      }
    }
    case 1: {
      return {
        [storage.section]: {
          [storage.method]: {
            [decodedKey.args[0].toString()]: decodeValue(),
          },
        },
      }
    }
    default:
      return {
        [storage.section]: {
          [storage.method]: decodeValue(),
        },
      }
  }
}

export const decodeStorageDiff = async (block: Block, diff: [HexString, HexString | null][]) => {
  const oldState = {}
  const newState = {}
  const meta = await block.meta
  for (const [key, value] of diff) {
    _.merge(oldState, decodeKeyValue(meta, block, key, (await block.get(key)) as any))
    _.merge(newState, decodeKeyValue(meta, block, key, value))
  }
  const oldStateWithoutEvents = _.cloneDeep(oldState)
  if (oldStateWithoutEvents['system']?.['events']) {
    oldStateWithoutEvents['system']['events'] = []
  }
  return [oldState, newState, diffPatcher.diff(oldStateWithoutEvents, newState)]
}
