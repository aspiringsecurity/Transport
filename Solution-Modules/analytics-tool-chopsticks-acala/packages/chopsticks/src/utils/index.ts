import { HexString } from '@polkadot/util/types'
import { StorageKey } from '@polkadot/types'
import { compactStripLength, hexToU8a, u8aToHex } from '@polkadot/util'

import { Blockchain } from '../blockchain'

export type GetKeys = (startKey?: string) => Promise<StorageKey<any>[]>

export type ProcessKey = (key: StorageKey<any>) => any

export async function fetchKeys(getKeys: GetKeys, processKey: ProcessKey) {
  const processKeys = async (keys: StorageKey<any>[]) => {
    for (const key of keys) {
      await processKey(key)
    }

    if (keys.length > 0) {
      return keys[keys.length - 1]
    }

    return undefined
  }

  const keys = await getKeys()
  let nextKey = await processKeys(keys)
  while (nextKey) {
    const keys = await getKeys(nextKey.toHex())
    nextKey = await processKeys(keys)
  }
}

export async function fetchKeysToArray(getKeys: GetKeys) {
  const res = [] as StorageKey<any>[]
  await fetchKeys(getKeys, (key) => res.push(key))
  return res
}

export const compactHex = (value: Uint8Array): HexString => {
  return u8aToHex(compactStripLength(value)[1])
}

export const getParaId = async (chain: Blockchain) => {
  const meta = await chain.head.meta
  const raw = await chain.head.get(compactHex(meta.query.parachainInfo.parachainId()))
  if (!raw) throw new Error('Cannot find parachain id')
  return meta.registry.createType('u32', hexToU8a(raw))
}

export const isUrl = (url: string) => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}

export type Deferred<T> = {
  resolve: (value: T | PromiseLike<T>) => void
  reject: (reason?: any) => void
  promise: Promise<T>
}

export function defer<T>() {
  const deferred = {} as Deferred<T>
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve
    deferred.reject = reject
  })
  return deferred
}
