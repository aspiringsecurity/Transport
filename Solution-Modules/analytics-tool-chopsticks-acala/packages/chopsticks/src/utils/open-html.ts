import { execSync } from 'node:child_process'

export const openHtml = (filePath: string) => {
  const start = process.platform == 'darwin' ? 'open' : process.platform == 'win32' ? 'start' : 'xdg-open'
  execSync(start + ' ' + filePath)
}
