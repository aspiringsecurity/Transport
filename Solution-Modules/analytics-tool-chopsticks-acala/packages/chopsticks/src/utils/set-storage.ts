import { DecoratedMeta } from '@polkadot/types/metadata/decorate/types'
import { HexString } from '@polkadot/util/types'
import { StorageKey } from '@polkadot/types'
import { stringCamelCase } from '@polkadot/util/string'
import { u8aToHex } from '@polkadot/util'

import { Blockchain } from '../blockchain'
import { StorageValueKind } from '../blockchain/storage-layer'

type RawStorageValues = [string, string | null][]
type StorageConfig = Record<string, Record<string, any>>
export type StorageValues = RawStorageValues | StorageConfig

function objectToStorageItems(meta: DecoratedMeta, storage: StorageConfig): RawStorageValues {
  const storageItems: RawStorageValues = []
  for (const sectionName in storage) {
    const section = storage[sectionName]

    const pallet = meta.query[stringCamelCase(sectionName)]
    if (!pallet) throw Error(`Cannot find pallet ${sectionName}`)

    for (const storageName in section) {
      const storage = section[storageName]

      if (storageName === '$removePrefix') {
        for (const mapName of storage) {
          const storageEntry = pallet[stringCamelCase(mapName)]
          if (!storageEntry) throw Error(`Cannot find storage ${mapName} in pallet ${sectionName}`)

          const prefix = storageEntry.keyPrefix()
          storageItems.push([u8aToHex(prefix), StorageValueKind.DeletedPrefix])
        }
        continue
      }

      const storageEntry = pallet[stringCamelCase(storageName)]
      if (!storageEntry) throw Error(`Cannot find storage ${storageName} in pallet ${sectionName}`)

      if (storageEntry.meta.type.isPlain) {
        const key = new StorageKey(meta.registry, [storageEntry])
        const type = storageEntry.meta.modifier.isOptional ? `Option<${key.outputType}>` : key.outputType
        storageItems.push([key.toHex(), storage ? meta.registry.createType(type, storage).toHex() : null])
      } else {
        for (const [keys, value] of storage) {
          const key = new StorageKey(meta.registry, [storageEntry, keys])
          const type = storageEntry.meta.modifier.isOptional ? `Option<${key.outputType}>` : key.outputType
          storageItems.push([key.toHex(), value ? meta.registry.createType(type, value).toHex() : null])
        }
      }
    }
  }
  return storageItems
}

export const setStorage = async (
  chain: Blockchain,
  storage: StorageValues,
  blockHash?: HexString
): Promise<HexString> => {
  const block = await chain.getBlock(blockHash)
  if (!block) throw Error(`Cannot find block ${blockHash || 'latest'}`)

  let storageItems: RawStorageValues
  if (Array.isArray(storage)) {
    storageItems = storage
  } else {
    storageItems = objectToStorageItems(await block.meta, storage)
  }
  block.pushStorageLayer().setAll(storageItems)
  return block.hash
}
