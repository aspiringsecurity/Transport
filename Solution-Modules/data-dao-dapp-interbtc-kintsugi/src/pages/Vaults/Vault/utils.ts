import { CollateralStatus, CollateralStatusRanges } from './types';

const getCollateralStatus = (value: number, ranges: CollateralStatusRanges, isInfinity: boolean): CollateralStatus => {
  if (isInfinity) return 'success';
  if (value <= ranges['error'].max) return 'error';
  if (value <= ranges['warning'].max) return 'warning';
  return 'success';
};

export { getCollateralStatus };
