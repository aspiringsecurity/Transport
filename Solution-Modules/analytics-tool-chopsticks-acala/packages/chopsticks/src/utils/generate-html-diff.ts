import { Block } from '../blockchain/block'
import { HexString } from '@polkadot/util/types'
import { decodeStorageDiff } from './decoder'
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import { template } from 'lodash'
import path from 'node:path'

export const generateHtmlDiff = async (block: Block, diff: [HexString, HexString | null][]) => {
  const [left, _right, delta] = await decodeStorageDiff(block, diff)
  const htmlTemplate = readFileSync(path.join(__dirname, '../../template/diff.html'), 'utf-8')
  return template(htmlTemplate)({ left: JSON.stringify(left), delta: JSON.stringify(delta) })
}

export const generateHtmlDiffPreviewFile = async (
  block: Block,
  diff: [HexString, HexString | null][],
  filename: string
) => {
  const html = await generateHtmlDiff(block, diff)
  mkdirSync('./preview', { recursive: true })
  const filePath = `./preview/${filename}.html`
  writeFileSync(filePath, html)
  return filePath
}
