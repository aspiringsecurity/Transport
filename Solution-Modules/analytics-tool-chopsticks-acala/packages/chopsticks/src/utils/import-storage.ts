import { HexString } from '@polkadot/util/types'
import { existsSync, readFileSync } from 'node:fs'
import yaml from 'js-yaml'

import { Blockchain } from '../blockchain'
import { StorageValues, setStorage } from './set-storage'
import { defaultLogger } from '../logger'

export const importStorage = async (chain: Blockchain, storage?: string | StorageValues, at?: HexString) => {
  if (storage == null) {
    return
  }
  let storageValue: StorageValues
  if (typeof storage === 'string') {
    if (!existsSync(storage)) throw Error(`File ${storage} does not exist`)
    storageValue = yaml.load(String(readFileSync(storage))) as StorageValues
  } else {
    storageValue = storage
  }
  const blockHash = await setStorage(chain, storageValue, at)
  defaultLogger.trace({ blockHash, storage }, 'ImportStorage')
}

export const overrideWasm = async (chain: Blockchain, wasmPath?: string, at?: HexString) => {
  if (wasmPath == null) {
    return
  }
  const wasm = readFileSync(wasmPath)
  let wasmHex: string
  if (wasm.at(0) === 0x30 && wasm.at(1) === 0x78) {
    // starts with 0x
    wasmHex = wasm.toString().trim()
  } else {
    wasmHex = '0x' + wasm.toString('hex')
  }
  if (at) {
    const block = await chain.getBlock(at)
    if (!block) throw new Error(`Cannot find block ${at}`)
    block.setWasm(wasmHex as HexString)
  } else {
    chain.head.setWasm(wasmHex as HexString)
  }
}
