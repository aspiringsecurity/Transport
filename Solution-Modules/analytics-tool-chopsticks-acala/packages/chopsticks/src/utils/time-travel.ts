import { BN, hexToU8a, u8aToHex } from '@polkadot/util'
import { HexString } from '@polkadot/util/types'
import { Slot } from '@polkadot/types/interfaces'

import { Blockchain } from '../blockchain'
import { compactHex } from '.'
import { getAuraSlotDuration } from '../executor'
import { setStorage } from './set-storage'

export const getCurrentSlot = async (chain: Blockchain) => {
  const meta = await chain.head.meta
  // use raw key here because some chain did not expose those storage to metadata
  const slotRaw = meta.consts.babe
    ? await chain.head.get('0x1cb6f36e027abb2091cfb5110ab5087f06155b3cd9a8c9e5e9a23fd5dc13a5ed') // babe.currentSlot
    : await chain.head.get('0x57f8dc2f5ab09467896f47300f04243806155b3cd9a8c9e5e9a23fd5dc13a5ed') // aura.currentSlot
  if (!slotRaw) throw new Error('Cannot find current slot')
  return meta.registry.createType<Slot>('Slot', hexToU8a(slotRaw)).toNumber()
}

export const getCurrentTimestamp = async (chain: Blockchain) => {
  const meta = await chain.head.meta
  const currentTimestampRaw = (await chain.head.get(compactHex(meta.query.timestamp.now()))) || '0x'
  return meta.registry.createType('u64', hexToU8a(currentTimestampRaw)).toNumber()
}

export const getSlotDuration = async (chain: Blockchain) => {
  const meta = await chain.head.meta
  return meta.consts.babe
    ? (meta.consts.babe.expectedBlockTime as any as BN).toNumber()
    : meta.query.aura
    ? getAuraSlotDuration(await chain.head.wasm, meta.registry)
    : 12_000
}

export const timeTravel = async (chain: Blockchain, timestamp: number) => {
  const meta = await chain.head.meta

  const slotDuration = await getSlotDuration(chain)
  const newSlot = Math.floor(timestamp / slotDuration)

  // new timestamp
  const storage: [HexString, HexString][] = [
    [compactHex(meta.query.timestamp.now()), u8aToHex(meta.registry.createType('u64', timestamp).toU8a())],
  ]

  if (meta.consts.babe) {
    // new slot
    storage.push([
      compactHex(meta.query.babe.currentSlot()),
      u8aToHex(meta.registry.createType('Slot', newSlot).toU8a()),
    ])

    // new epoch
    const epochDuration = (meta.consts.babe.epochDuration as any as BN).toNumber()
    const newEpoch = Math.floor(timestamp / epochDuration)
    storage.push([
      compactHex(meta.query.babe.epochIndex()),
      u8aToHex(meta.registry.createType('u64', newEpoch).toU8a()),
    ])
  } else if (meta.query.aura) {
    // new slot
    storage.push([
      compactHex(meta.query.aura.currentSlot()),
      u8aToHex(meta.registry.createType('Slot', newSlot).toU8a()),
    ])
  }

  await setStorage(chain, storage)
}
