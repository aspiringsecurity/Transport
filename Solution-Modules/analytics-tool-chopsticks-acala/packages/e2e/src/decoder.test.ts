import { afterAll, describe, expect, it } from 'vitest'
import { decodeKey, decodeKeyValue } from '@acala-network/chopsticks/utils/decoder'

import networks from './networks'

const SYSTEM_ACCOUNT =
  '0x26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9de1e86a9a8c739864cf3cc5ec2bea59fd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d'
const TOKENS_ACCOUNTS =
  '0x99971b5749ac43e0235e41b0d37869188ee7418a6531173d60d1f6a82d8f4d51de1e86a9a8c739864cf3cc5ec2bea59fd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d01a12dfa1fa4ab9a0000'

describe('decoder', async () => {
  const acala = await networks.acala()
  const { chain } = acala

  afterAll(async () => {
    await acala.teardown()
  })

  it('decode keys', async () => {
    const { storage, decodedKey } = decodeKey(await chain.head.meta, chain.head, SYSTEM_ACCOUNT)
    expect(storage?.section).eq('system')
    expect(storage?.method).eq('account')
    expect(decodedKey?.args.map((x) => x.toHuman())).contains('25fqepuLngYL2DK9ApTejNzqPadUUZ9ALYyKWX2jyvEiuZLa')
  })

  it('decode key-value', async () => {
    const meta = await chain.head.meta
    const data = { data: { free: 10000000000 } }
    const value = meta.registry.createType('AccountInfo', data)
    expect(decodeKeyValue(meta, chain.head, SYSTEM_ACCOUNT, value.toHex())).toMatchSnapshot()

    const ormlAccountData = meta.registry.createType('AccountData', data.data)
    expect(decodeKeyValue(meta, chain.head, TOKENS_ACCOUNTS, ormlAccountData.toHex())).toMatchSnapshot()
    await new Promise((resolve) => setTimeout(resolve, 1000))
  })

  it('works with multiple chains', async () => {
    const { chain, teardown } = await networks.polkadot()

    const meta = await chain.head.meta
    const data = { data: { free: 10000000000 } }
    const value = meta.registry.createType('AccountInfo', data)
    expect(decodeKeyValue(meta, chain.head, SYSTEM_ACCOUNT, value.toHex())).toMatchSnapshot()

    await teardown()
  })
})
