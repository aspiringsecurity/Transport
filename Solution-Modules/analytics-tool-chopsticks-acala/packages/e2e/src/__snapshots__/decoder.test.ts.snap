// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`decoder > decode key-value 1`] = `
{
  "system": {
    "account": {
      "25fqepuLngYL2DK9ApTejNzqPadUUZ9ALYyKWX2jyvEiuZLa": {
        "consumers": "0",
        "data": {
          "feeFrozen": "0",
          "free": "10,000,000,000",
          "miscFrozen": "0",
          "reserved": "0",
        },
        "nonce": "0",
        "providers": "0",
        "sufficients": "0",
      },
    },
  },
}
`;

exports[`decoder > decode key-value 2`] = `
{
  "tokens": {
    "accounts": {
      "25fqepuLngYL2DK9ApTejNzqPadUUZ9ALYyKWX2jyvEiuZLa": {
        "{\\"token\\":\\"ACA\\"}": {
          "free": "10,000,000,000",
          "frozen": "0",
          "reserved": "0",
        },
      },
    },
  },
}
`;

exports[`decoder > works with multiple chains 1`] = `
{
  "system": {
    "account": {
      "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5": {
        "consumers": "0",
        "data": {
          "feeFrozen": "0",
          "free": "10,000,000,000",
          "miscFrozen": "0",
          "reserved": "0",
        },
        "nonce": "0",
        "providers": "0",
        "sufficients": "0",
      },
    },
  },
}
`;
